File: examples/basic.c
/* SPDX-License-Identifier: Apache-2.0 */
/*
** Basic KeyValueStore Examples
** Demonstrates: Hello World, CRUD operations, existence checks
*/

#define SNKV_IMPLEMENTATION
#include "snkv.h"
#include <stdio.h>
#include <string.h>

static void print_user(const char *user_id, KeyValueStore *pKV) {
    void *pValue;
    int nValue;

    int rc = keyvaluestore_get(pKV, user_id, strlen(user_id), &pValue, &nValue);
    if (rc == KEYVALUESTORE_OK) {
        printf("User %s: %.*s\n", user_id, nValue, (char*)pValue);
        snkv_free(pValue);
    } else if (rc == KEYVALUESTORE_NOTFOUND) {
        printf("User %s: Not found\n", user_id);
    }
}

static void example_hello_world(void) {
    KeyValueStore *pKV = NULL;
    int rc;

    printf("=== Hello World ===\n");

    rc = keyvaluestore_open("hello.db", &pKV, KEYVALUESTORE_JOURNAL_WAL);
    if (rc != KEYVALUESTORE_OK) {
        fprintf(stderr, "Failed to open database\n");
        return;
    }

    /* Store a greeting */
    const char *key = "greeting";
    const char *value = "Hello, World!";
    rc = keyvaluestore_put(pKV, key, strlen(key), value, strlen(value));
    if (rc == KEYVALUESTORE_OK) {
        printf("Stored: %s = %s\n", key, value);
    }

    /* Retrieve the greeting */
    void *pValue = NULL;
    int nValue = 0;
    rc = keyvaluestore_get(pKV, key, strlen(key), &pValue, &nValue);
    if (rc == KEYVALUESTORE_OK) {
        printf("Retrieved: %s = %.*s\n", key, nValue, (char*)pValue);
        snkv_free(pValue);
    }

    keyvaluestore_close(pKV);
    remove("hello.db");
    printf("\n");
}

static void example_crud(void) {
    KeyValueStore *pKV;

    printf("=== CRUD Operations ===\n");

    keyvaluestore_open("users.db", &pKV, KEYVALUESTORE_JOURNAL_WAL);

    /* CREATE */
    printf("--- CREATE ---\n");
    keyvaluestore_put(pKV, "user:1", 6, "Alice Smith", 11);
    print_user("user:1", pKV);

    /* READ */
    printf("--- READ ---\n");
    print_user("user:1", pKV);

    /* UPDATE */
    printf("--- UPDATE ---\n");
    keyvaluestore_put(pKV, "user:1", 6, "Alice Johnson", 13);
    print_user("user:1", pKV);

    /* DELETE */
    printf("--- DELETE ---\n");
    keyvaluestore_delete(pKV, "user:1", 6);
    print_user("user:1", pKV);

    keyvaluestore_close(pKV);
    remove("users.db");
    printf("\n");
}

static void example_existence(void) {
    KeyValueStore *pKV;

    printf("=== Existence Check ===\n");

    keyvaluestore_open("inventory.db", &pKV, KEYVALUESTORE_JOURNAL_WAL);

    keyvaluestore_put(pKV, "item:laptop", 11, "In Stock", 8);
    keyvaluestore_put(pKV, "item:mouse", 10, "Out of Stock", 12);

    const char *items[] = {"item:laptop", "item:mouse", "item:keyboard"};
    int num_items = 3;

    for (int i = 0; i < num_items; i++) {
        int exists = 0;
        keyvaluestore_exists(pKV, items[i], strlen(items[i]), &exists);
        printf("%s: %s\n", items[i], exists ? "EXISTS" : "NOT FOUND");
    }

    keyvaluestore_close(pKV);
    remove("inventory.db");
    printf("\n");
}

int main(void) {
    example_hello_world();
    example_crud();
    example_existence();
    printf("All basic examples passed.\n");
    return 0;
}

File: python/examples/basic.py
# SPDX-License-Identifier: Apache-2.0
"""
Basic SNKV Examples
Demonstrates: Hello World, CRUD operations, existence checks

Run:
    python examples/basic.py
"""

import os
import snkv
from snkv import Store, NotFoundError

DB_FILE = "basic_example.db"


def hello_world():
    print("--- Hello World ---")
    with Store(DB_FILE) as db:
        db.put(b"greeting", b"Hello, SNKV!")

        value = db.get(b"greeting")
        print(f"greeting = {value.decode()}")


def crud_operations():
    print("\n--- CRUD Operations ---")
    with Store(DB_FILE) as db:
        # Create
        db["user:1"] = "Alice"
        db["user:2"] = "Bob"
        db["user:3"] = "Charlie"
        print("Inserted 3 users")

        # Read
        print(f"user:1 = {db['user:1'].decode()}")
        print(f"user:2 = {db['user:2'].decode()}")

        # Update
        db["user:1"] = "Alice Smith"
        print(f"user:1 updated = {db['user:1'].decode()}")

        # Delete
        del db["user:3"]
        print("user:3 deleted")

        # Confirm deletion via get with default
        result = db.get("user:3", default=b"(not found)")
        print(f"user:3 after delete = {result.decode()}")


def existence_checks():
    print("\n--- Existence Checks ---")
    with Store(DB_FILE) as db:
        db["present"] = "yes"

        print(f"'present' exists: {db.exists('present')}")
        print(f"'missing' exists: {db.exists('missing')}")
        print(f"'present' in db:  {'present' in db}")
        print(f"'missing' in db:  {'missing' in db}")


def error_handling():
    print("\n--- Error Handling ---")
    with Store(DB_FILE) as db:
        db["key"] = "value"

        # KeyError on missing key via [] syntax
        try:
            _ = db["does_not_exist"]
        except KeyError as e:
            print(f"KeyError caught: {e}")

        # NotFoundError is also raised (it inherits from KeyError)
        try:
            _ = db["does_not_exist"]
        except NotFoundError:
            print("NotFoundError caught (same exception)")

        # get() with default avoids exceptions
        val = db.get("does_not_exist", b"fallback")
        print(f"get with default: {val.decode()}")


def binary_data():
    print("\n--- Binary Keys and Values ---")
    with Store(DB_FILE) as db:
        # Keys and values can be arbitrary binary data
        binary_key   = bytes([0x00, 0x01, 0xFF])
        binary_value = bytes(range(256))

        db.put(binary_key, binary_value)
        retrieved = db.get(binary_key)
        print(f"Binary round-trip OK: {retrieved == binary_value}")
        print(f"Value length: {len(retrieved)} bytes")


def in_memory_store():
    print("\n--- In-Memory Store ---")
    # Pass None for path to use an in-memory database (no file created)
    with Store(None) as db:
        db["temp"] = "ephemeral data"
        print(f"In-memory value: {db['temp'].decode()}")
    print("Store closed - data is gone (no file was created)")


if __name__ == "__main__":
    hello_world()
    crud_operations()
    existence_checks()
    error_handling()
    binary_data()
    in_memory_store()

    # Cleanup
    for ext in ("", "-wal", "-shm"):
        try:
            os.remove(DB_FILE + ext)
        except FileNotFoundError:
            pass
    print("\n[OK] basic.py complete")
